### ISSUE TITLE
RandomAccessFile setPosition not working as expected 

### ISSUE THREAD

**Issue by Henco1** [dart-lang/sdk#45238]
I am trying to chunk a file into base 64 string parts but I am having an issue with streams read incorrectly & the length of the returned strings.
```
File file = File(result.files.single.path);
RandomAccessFile randomAccess = file.openSync(mode: FileMode.read);

randomAccess = await randomAccess.setPosition(0);
List<int> buff = randomAccess.readSync(10);
debugPrint(base64Encode(buff)); // outputs: /9j/4AAQSkZJRg==

randomAccess = await randomAccess.setPosition(10);
List<int> buff2 = randomAccess.readSync(10);
debugPrint(base64Encode(buff2)); // outputs: AAEBAAABAAEAAA==

// Correct string
// /9j/4AAQSkZJRgABAQAAAQABAAD//gA7Q1JFQVRP...
```
The first string returned is correct but the length does not match, 14 characters are returned for the first base 64 string, I expected 10, one byte for each character

The seconds string is also 14 characters and is nowhere to be found in the complete base64 file string (which i obtained by converting the entire file) i.e. if I convert the entire file at once I cannot find a pattern matching buff2's base64 segment

I used [this](https://macaulay.cuny.edu/seminars/ebin07/images/thumb/1/14/TEST.IMAGE.jpg/320px-TEST.IMAGE.jpg) testing image.

```
Flutter 2.0.1 • channel stable • https://github.com/flutter/flutter.git
Framework • revision c5a4b4029c (4 days ago) • 2021-03-04 09:47:48 -0800
Engine • revision 40441def69
Tools • Dart 2.12.0
```

flutter doctor
```
Doctor summary (to see all details, run flutter doctor -v):
[✓] Flutter (Channel stable, 2.0.1, on Mac OS X 10.15.7 19H114 darwin-x64, locale en-ZA)
[✓] Android toolchain - develop for Android devices (Android SDK version 29.0.2)
[✓] Xcode - develop for iOS and macOS
[✓] Chrome - develop for the web
[✓] Android Studio (version 4.1)
[✓] VS Code (version 1.54.1)
[✓] Connected device (2 available)

• No issues found!
```

**Comment by sortie** [comment]
base64 encodes 6 bits of input per 8 bits of output data. That means the output might not be aligned if the input bits don't divide by six. base64 pads the output with characters like = to align the output in this case. That is why the first encoded fragment has two extra = characters.

Since the first fragment needed alignment padding, that means the base64 encoder would have mixed data from the first and second data into the 'A' character following '/9j/4AAQSkZJRg'. But by encoding the second fragment separately, you force the encoder to reset and not combine the encoding of the fragments. The result is that get another string.

In short, while base64 data can be concatenated, you can't split the data into two parts and encode them separately and expect to get the same as encoding them together.

If that doesn't explain your problem, then please reply on the bug with a reproduction case where you show the actual integer byte values read and what you expected to read, so we know there really is a RandomAccessFile setPosition bug.

**Comment by Henco1** [comment]
Thanks for the explanation, could you please add this description to the documentation?

https://api.dart.dev/stable/2.12.0/dart-io/RandomAccessFile/readSync.html
https://api.dart.dev/stable/2.10.5/dart-convert/base64Encode.html

**Comment by Henco1** [comment]
Also see [43577](https://github.com/dart-lang/sdk/issues/43577)

**Comment by sortie** [comment]
The readSync documentation is already correct and says it reads up to the number of bytes requested.

The base64 documentation links to the base64 specification https://tools.ietf.org/html/rfc4648 that explains how base64 works. Potentially that documentation could gain a description of what base64 works although that can also be looked up at other sources.
